zkevm-circuits/src/ecc_circuit.rs

```diff
@@ -674,43 +698,177 @@ impl<F: Field, const XI_0: i64> EccCircuit<F, XI_0> {
+            point_r: G1Assigned {
+                x_rlc: ecc_chip.field_chip().range().gate().inner_product(
+                    ctx,
+                    ec_mul_decomposed.point_r.x_cells.clone(),
+                    keccak_powers.iter().cloned(),
+                ),
+                y_rlc: ecc_chip.field_chip().range().gate().inner_product(
+                    ctx,
+                    ec_mul_decomposed.point_r.y_cells.clone(),
+                    keccak_powers.iter().cloned(),
+                ),
+            },
+        }
+    }
+
+    /// Handles Phase2 for EcPairing operation and returns the RLC'd input bytes.
+    fn assign_ec_pairing(
+        &self,
+        ctx: &mut Context<F>,
+        ec_pairing_decomposed: &EcPairingDecomposed<F>,
+        ecc_chip: &EccChip<F, FpConfig<F, Fq>>,
+        keccak_powers: &[QuantumCell<F>],
+    ) -> EcPairingAssigned<F> {
+        EcPairingAssigned {
+            input_rlc: ecc_chip.field_chip().range().gate().inner_product(
+                ctx,
+                ec_pairing_decomposed.input_cells.clone().into_iter().rev(),
+                keccak_powers.iter().cloned(),
+            ),
+            success: ec_pairing_decomposed.success,
+        }
+    }
+
+    /// Assign the G1 EcPoint and return its decomposed state.
     fn assign_g1(
         &self,
         ctx: &mut Context<F>,
-        fp_chip: &EccChip<F, FpConfig<F, Fq>>,
+        ecc_chip: &EccChip<F, FpConfig<F, Fq>>,
         g1: G1Affine,
-        powers_of_rand: &[QuantumCell<F>],
-    ) -> G1Assigned<F> {
-        let ec_point = fp_chip.load_private(ctx, (Value::known(g1.x), Value::known(g1.y)));
+    ) -> G1Decomposed<F> {
+        let ec_point = ecc_chip.load_private(ctx, (Value::known(g1.x), Value::known(g1.y)));
         let (x_cells, y_cells) = self.decompose_g1(g1);
-        let decomposed = G1Decomposed {
+        let powers_of_256 = iter::successors(Some(F::one()), |coeff| Some(F::from(256) * coeff))
+            .take(32)
+            .map(|x| QuantumCell::Constant(x))
+            .collect_vec();
+        self.assert_crt_repr(ctx, ecc_chip, &ec_point.x, &x_cells, &powers_of_256);
+        self.assert_crt_repr(ctx, ecc_chip, &ec_point.y, &y_cells, &powers_of_256);
+        G1Decomposed {
             ec_point,
-            x_cells: x_cells.clone(),
-            y_cells: x_cells.clone(),
-        };
-        G1Assigned {
-            decomposed,
-            x_rlc: fp_chip.field_chip().range.gate.inner_product(
+            x_cells,
+            y_cells,
+        }
+    }
+
+    /// Assert that a CRT integer's bytes representation is correct.
+    fn assert_crt_repr(
+        &self,
+        ctx: &mut Context<F>,
+        ecc_chip: &EccChip<F, FpConfig<F, Fq>>,
+        crt_int: &CRTInteger<F>,
+        bytes: &[QuantumCell<F>],
+        powers_of_256: &[QuantumCell<F>],
+    ) {
+        debug_assert_eq!(bytes.len(), 32);
+        debug_assert!(powers_of_256.len() >= 11);
+
+        let [limb1, limb2, limb3] = [
```

Comment:
nits: seems like you can directly name it as `limbs`? it doesn't look like you need access to individual elements